// hyperdispatch/lib/codegen.js
const s = require('generate-string')

module.exports = function generateCode(hyperdispatch, { esm = false }) {
  let str = ''
  str += '// This file is autogenerated by the hyperdispatch compiler\n'
  str += '/* eslint-disable camelcase */\n'
  str += '\n'

  if (esm) {
    str += "import { c, b4a, assert } from 'hyperdispatch/runtime'\n"
    str += "import { version, getEncoding, setVersion } from './messages.js'\n"
  } else {
    str += "const { c, b4a, assert } = require('hyperdispatch/runtime')\n"
    str += "const { version, getEncoding, setVersion } = require('./messages.js')\n"
  }

  str += '\n'
  str += 'const defaultVersion = version\n'
  str += '\n'

  str += 'class Router {\n'
  str += '  constructor () {\n'
  for (const handler of hyperdispatch.handlers) {
    str += `    this._handler${handler.id} = null\n`
  }
  str += '\n'
  str += `    this._missing = ${hyperdispatch.handlers.length}\n`
  str += '  }\n'
  str += '\n'
  str += '  add (name, handler) {\n'
  str += '    switch (name) {\n'
  for (const handler of hyperdispatch.handlers) {
    str += `      case ${s(handler.name)}:\n`
    str += `        this._handler${handler.id} = handler\n`
    str += '        break\n'
  }
  str += '      default:\n'
  str += "        throw new Error('Cannot register a handler for a nonexistent route: ' + name)\n"
  str += '    }\n'
  str += '    this._missing--\n'
  str += '  }\n'
  str += '\n'
  str += '  _checkAll () {\n'
  for (const handler of hyperdispatch.handlers) {
    str += `    assert(this._handler${handler.id} !== null, 'Missing handler for ${JSON.stringify(handler.name)}')\n`
  }
  str += '  }\n'
  str += '\n'
  str += '  async dispatch (message, context) {\n'
  str += '    if (this._missing > 0) {\n'
  str += '      this._checkAll()\n'
  str += '    }\n'
  str += '\n'
  str += '    setVersion(defaultVersion)\n'
  str += '\n'
  str += '    const op = b4a.isBuffer(message) ? decode(message) : message\n'
  str += '\n'
  str += '    switch (op.id) {\n'
  for (const handler of hyperdispatch.handlers) {
    str += `      case ${handler.id}:\n`
    str += `        return this._handler${handler.id}(op.value, context)\n`
  }
  str += '      default:\n'
  str += "        throw new Error('Handler not found for ID:' + op.id)\n"
  str += '    }\n'
  str += '  }\n'
  str += '}\n'

  str += '\n'

  str += 'function encode (name, message, { version = defaultVersion } = {}) {\n'
  str += '  const state = { buffer: null, start: 0, end: 0 }\n'
  str += '\n'
  str += '  const route = getRouteByName(name)\n'
  str += '  setVersion(version)\n'
  str += '\n'
  str += '  c.uint.preencode(state, route.id)\n'
  str += '  route.enc.preencode(state, message)\n'
  str += '\n'
  str += '  state.buffer = b4a.allocUnsafe(state.end)\n'
  str += '  c.uint.encode(state, route.id)\n'
  str += '  route.enc.encode(state, message)\n'
  str += '\n'
  str += '  return state.buffer\n'
  str += '}\n'
  str += '\n'

  str += 'function decode (buffer, { version = defaultVersion } = {}) {\n'
  str += '  const state = { buffer, start: 0, end: buffer.length }\n'
  str += '\n'
  str += '  const id = c.uint.decode(state)\n'
  str += '  const route = getRouteById(id)\n'
  str += '  setVersion(version)\n'
  str += '\n'
  str += '  const value = route.enc.decode(state)'
  str += '\n'
  str += '  return { id, name: route.name, value }\n'
  str += '}\n'
  str += '\n'

  for (const handler of hyperdispatch.handlers) {
    str += 'const route' + handler.id + ' = {\n'
    str += `  name: ${s(handler.name)},\n`
    str += `  id: ${handler.id},\n`
    str += `  enc: getEncoding(${s(handler.requestType)})\n`
    str += '}\n'
    str += '\n'
  }

  str += 'function getRouteByName (name) {\n'
  str += '  switch (name) {\n'
  for (const handler of hyperdispatch.handlers) {
    str += `    case ${s(handler.name)}:\n`
    str += `      return route${handler.id}\n`
  }
  str += '    default:\n'
  str += "      throw new Error('Handler not found for name: ' + name)\n"
  str += '  }\n'
  str += '}\n'
  str += '\n'

  str += 'function getRouteById (id) {\n'
  str += '  switch (id) {\n'
  for (const handler of hyperdispatch.handlers) {
    str += `    case ${handler.id}:\n`
    str += `      return route${handler.id}\n`
  }
  str += '    default:\n'
  str += "      throw new Error('Handler not found for ID: ' + id)\n"
  str += '  }\n'
  str += '}\n'
  str += '\n'

  if (esm) {
    str += 'export {\n'
    str += '  version,\n'
    str += '  encode,\n'
    str += '  decode,\n'
    str += '  Router\n'
    str += '}\n'
  } else {
    str += 'module.exports = {\n'
    str += '  version,\n'
    str += '  encode,\n'
    str += '  decode,\n'
    str += '  Router\n'
    str += '}\n'
  }

  return str
}
